app/lib/shared/integrity.dart

```diff
@@ -602,28 +595,54 @@ class IntegrityChecker {
   }
 
   Stream<String> _checkTarballInBuckets(
-      PackageVersion pv, Uri archiveDownloadUri) async* {
-    final info =
-        await packageBackend.packageTarballInfo(pv.package, pv.version!);
-    if (info == null) {
-      yield 'PackageVersion "${pv.qualifiedVersionKey}" has no matching archive file.';
-    }
+    PackageVersion pv,
+    Uri archiveDownloadUri, {
+    required bool shouldBeInPublicBucket,
+  }) async* {
     final canonicalInfo = await storageService
         .bucket(activeConfiguration.canonicalPackagesBucketName!)
         // ignore: invalid_use_of_visible_for_testing_member
         .tryInfo(tarballObjectName(pv.package, pv.version!));
+    if (canonicalInfo == null) {
+      yield 'PackageVersion "${pv.qualifiedVersionKey}" has no matching canonical archive file.';
+      return;
+    }
 
-    if (canonicalInfo != null) {
-      if (!canonicalInfo.hasSameSignatureAs(info)) {
-        yield 'Canonical archive for PackageVersion "${pv.qualifiedVersionKey}" differs in old bucket.';
+    final info =
+        await packageBackend.packageTarballInfo(pv.package, pv.version!);
+    if (info == null) {
+      if (shouldBeInPublicBucket) {
+        yield 'PackageVersion "${pv.qualifiedVersionKey}" has no matching public archive file.';
       }
+      return;
+    }
+
+    if (!shouldBeInPublicBucket) {
+      yield 'PackageVersion "${pv.qualifiedVersionKey}" has matching public archive file but it must not.';
+      return;
+    }
+
+    if (!canonicalInfo.hasSameSignatureAs(info)) {
+      yield 'Canonical archive for PackageVersion "${pv.qualifiedVersionKey}" differs in old bucket.';
```

Comment:
What is "old bucket"? Should this be:
```suggestion
      yield 'Canonical archive for PackageVersion "${pv.qualifiedVersionKey}" differs from public bucket.';
```